File: tools/test-pp-server.py
#!/usr/bin/env python3
"""Dummy PersonaPlex server to capture and save received Opus audio.
Usage: python3 tools/test-pp-server.py [--port PORT] [--out-dir DIR]

Start this, then open the frontend with:
  http://localhost:3000/?personaplex_ws=ws://localhost:8001/api/chat
"""

import asyncio
import json
import os
import struct
import sys
import time
from pathlib import Path

OUT_DIR = Path(
    sys.argv[sys.argv.index("--out-dir") + 1]
    if "--out-dir" in sys.argv
    else "tools/test-recordings"
)
PORT = int(sys.argv[sys.argv.index("--port") + 1]) if "--port" in sys.argv else 8001

OUT_DIR.mkdir(parents=True, exist_ok=True)


async def handle(websocket):
    """Handle a single PersonaPlex client connection."""
    session_id = time.strftime("%Y%m%d_%H%M%S")
    raw_path = OUT_DIR / f"session_{session_id}_opus.raw"
    print(f"\n[SERVER] Connection from {websocket.remote_address}")

    # Extract query params
    path = websocket.request.path if hasattr(websocket, "request") else "/"
    if "?" in path:
        params = dict(p.split("=") for p in path.split("?")[1].split("&") if "=" in p)
        print(f"[SERVER] Params: {params}")

    # Send PersonaPlex handshake (tag 0x00 followed by any data)
    handshake = bytearray([0])
    handshake.extend(b"ready")
    await websocket.send(bytes(handshake))
    print(f"[SERVER] Sent handshake")

    # Receive audio packets
    frames = []
    text_parts = []
    try:
        async for message in websocket:
            tag = message[0]
            payload = message[1:]
            if tag == 1:
                frames.append(bytes(payload))
                if len(frames) % 50 == 0:
                    print(f"[SERVER] Received {len(frames)} audio frames")
            elif tag == 2:
                text_parts.append(payload.decode("utf-8", errors="replace"))
                print(f"[SERVER] Text: {text_parts[-1][:60]}...")
    except Exception as e:
        print(f"[SERVER] Connection closed: {e}")

    # Save raw Opus data
    all_data = b"".join(frames)
    raw_path.write_bytes(all_data)
    print(f"[SERVER] Saved {len(frames)} frames ({len(all_data)} bytes) → {raw_path}")

    # Try to decode to WAV (requires opuslib or ffmpeg)
    wav_path = raw_path.with_suffix(".wav")
    try:
        import subprocess

        # Use ffmpeg to convert Opus → WAV
        subprocess.run(
            [
                "ffmpeg",
                "-y",
                "-f",
                "opus",
                "-i",
                str(raw_path),
                "-ar",
                "16000",
                "-ac",
                "1",
                str(wav_path),
            ],
            capture_output=True,
            timeout=10,
        )
        if wav_path.exists() and wav_path.stat().st_size > 100:
            print(f"[SERVER] Decoded WAV: {wav_path} ({wav_path.stat().st_size} bytes)")
    except Exception:
        print(f"[SERVER] ffmpeg not available — raw Opus saved to {raw_path}")

    if text_parts:
        txt_path = raw_path.with_suffix(".txt")
        txt_path.write_text("\n".join(text_parts))
        print(f"[SERVER] Transcript saved: {txt_path}")


async def main():
    try:
        import websockets
    except ImportError:
        print("Install websockets: pip install websockets")
        sys.exit(1)

    print(f"\n=== Dummy PersonaPlex Server ===")
    print(f"Listening on ws://0.0.0.0:{PORT}/api/chat")
    print(f"Saving recordings to: {OUT_DIR}/")
    print(f"\nUse this URL in the frontend:")
    print(f"  http://localhost:3000/?personaplex_ws=ws://localhost:{PORT}/api/chat")
    print()

    async with websockets.serve(handle, "0.0.0.0", PORT):
        await asyncio.Future()


if __name__ == "__main__":
    asyncio.run(main())

File: tools/test-pp-server.js
// Dummy PersonaPlex server for testing — saves received Opus audio
// Usage: node tools/test-pp-server.mjs [--port PORT]

const { createServer } = require("http");
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const PORT = parseInt(process.argv[process.argv.indexOf("--port") + 1]) || 8001;
const OUT_DIR = path.join(__dirname, "test-recordings");
if (!fs.existsSync(OUT_DIR)) fs.mkdirSync(OUT_DIR, { recursive: true });

const server = createServer((req, res) => {
  res.writeHead(200, { "Content-Type": "text/plain" });
  res.end("Dummy PersonaPlex — test server");
});

server.on("upgrade", (req, socket, head) => {
  const key = req.headers["sec-websocket-key"];
  if (!key) { socket.destroy(); return; }
  const accept = crypto.createHash("sha1").update(key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").digest("base64");
  socket.write("HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: " + accept + "\r\n\r\n");

  const ts = new Date().toISOString().replace(/[:.T-]/g, "_").slice(0, 15);
  const rawFile = path.join(OUT_DIR, `session_${ts}_opus.raw`);
  const frames = [];

  // Send handshake (tag 0)
  const hs = Buffer.alloc(6);
  hs[0] = 0; hs.write("ready", 1);
  sendFrame(socket, hs);
  console.log(`[SERVER] Connection from ${req.socket.remoteAddress}, sent handshake`);

  socket.on("data", (buf) => {
    try { const msg = decodeFrame(buf); if (msg) {
      if (msg === "CLOSE") { socket.end(); return; }
      if (msg[0] === 1) { frames.push(msg.slice(1)); if (frames.length % 50 === 0) console.log(`[SERVER] ${frames.length} audio frames`); }
      else if (msg[0] === 2) console.log(`[SERVER] Text: ${Buffer.from(msg.slice(1)).toString().slice(0, 60)}...`);
    }} catch {}
  });

  function close() {
    const raw = Buffer.concat(frames);
    fs.writeFileSync(rawFile, raw);
    console.log(`[SERVER] Saved ${frames.length} frames (${raw.length} bytes) → ${rawFile}`);
    // Auto-convert to WAV
    const wavFile = rawFile.replace(".raw", ".wav");
    const { execSync } = require("child_process");
    try {
      execSync(`ffmpeg -y -f ogg -i "${rawFile}" -ar 16000 -ac 1 "${wavFile}" 2>/dev/null`);
      console.log(`[SERVER] WAV: ${wavFile}`);
    } catch {}
  }
  socket.on("end", close);
  socket.on("close", close);
  socket.on("error", () => { try { close(); } catch {} });
});

function sendFrame(socket, payload) {
  const buf = Buffer.alloc(2 + payload.length);
  buf[0] = 0x82; // text frame, final
  buf[1] = payload.length;
  payload.copy(buf, 2);
  socket.write(buf);
}

function decodeFrame(data) {
  const opcode = data[0] & 0x0f;
  if (opcode === 0x8) return "CLOSE";
  if (opcode === 0x9) return null; // ping
  const masked = (data[1] & 0x80) !== 0;
  let len = data[1] & 0x7f;
  let off = 2;
  if (len === 126) { len = data.readUInt16BE(2); off = 4; }
  else if (len === 127) { len = Number(data.readBigUInt64BE(2)); off = 10; }
  if (!masked) return data.slice(off, off + len);
  const mask = data.slice(off, off + 4);
  const payload = Buffer.alloc(len);
  for (let i = 0; i < len; i++) payload[i] = data[off + 4 + i] ^ mask[i % 4];
  return payload;
}


server.listen(PORT, () => {
  console.log(`\n=== Dummy PersonaPlex Server ===`);
  console.log(`Listening on ws://0.0.0.0:${PORT}/api/chat`);
  console.log(`Saving recordings to: ${OUT_DIR}/`);
  console.log(`\nOpen: http://localhost:3000/?personaplex_ws=ws://localhost:${PORT}/api/chat`);
  console.log();
});